File: example/mission.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "cmavsdk/mavsdk.h"
#include "cmavsdk/plugins/action/action.h"
#include "cmavsdk/plugins/mission/mission.h"
#include "cmavsdk/plugins/telemetry/telemetry.h"

mavsdk_mission_mission_item_t make_mission_item(
    double latitude_deg,
    double longitude_deg,
    float relative_altitude_m,
    float speed_m_s,
    bool is_fly_through,
    float gimbal_pitch_deg,
    float gimbal_yaw_deg,
    mavsdk_mission_mission_item_camera_action_t camera_action)
{
    mavsdk_mission_mission_item_t item = {0};
    item.latitude_deg = latitude_deg;
    item.longitude_deg = longitude_deg;
    item.relative_altitude_m = relative_altitude_m;
    item.speed_m_s = speed_m_s;
    item.is_fly_through = is_fly_through;
    item.gimbal_pitch_deg = gimbal_pitch_deg;
    item.gimbal_yaw_deg = gimbal_yaw_deg;
    item.camera_action = camera_action;
    item.loiter_time_s = 0.0f;
    item.camera_photo_interval_s = 1.0;
    item.acceptance_radius_m = 1.0f;
    item.yaw_deg = 0.0f;
    item.camera_photo_distance_m = 0.0f;
    item.vehicle_action = MAVSDK_MISSION_MISSION_ITEM_VEHICLE_ACTION_NONE;
    return item;
}

void usage(const char* bin_name)
{
    fprintf(stderr, "Usage : %s <connection_url>\n", bin_name);
    fprintf(stderr, "Connection URL format should be :\n");
    fprintf(stderr, " For TCP server: tcpin://<our_ip>:<port>\n");
    fprintf(stderr, " For TCP client: tcpout://<remote_ip>:<port>\n");
    fprintf(stderr, " For UDP server: udpin://<our_ip>:<port>\n");
    fprintf(stderr, " For UDP client: udpout://<remote_ip>:<port>\n");
    fprintf(stderr, " For Serial : serial://</path/to/serial/dev>:<baudrate>]\n");
    fprintf(stderr, "For example, to connect to the simulator use URL: udpin://0.0.0.0:14540\n");
}

static volatile bool want_to_pause = false;

static void mission_progress_callback(
    const mavsdk_mission_mission_progress_t mission_progress,
    void* user_data)
{
    (void)user_data; // Unused
    
    printf("Mission status update: %d / %d\n",
           mission_progress.current,
           mission_progress.total);
    
    if (mission_progress.current >= 2) {
        want_to_pause = true;
    }
}

int main(int argc, char** argv)
{
    if (argc != 2) {
        usage(argv[0]);
        return 1;
    }
    
    // Create MAVSDK instance
    mavsdk_configuration_t config = 
        mavsdk_configuration_create_with_component_type(MAVSDK_COMPONENT_TYPE_GROUND_STATION);
    mavsdk_t mavsdk = mavsdk_create(config);
    
    // Add connection
    mavsdk_connection_result_t connection_result = 
        mavsdk_add_any_connection(mavsdk, argv[1]);
    
    if (connection_result != MAVSDK_CONNECTION_RESULT_SUCCESS) {
        fprintf(stderr, "Connection failed: %d\n", connection_result);
        mavsdk_destroy(mavsdk);
        return 1;
    }
    
    printf("Waiting for system...\n");
    mavsdk_system_t system = mavsdk_first_autopilot(mavsdk, 3.0);
    if (!system) {
        fprintf(stderr, "Timed out waiting for system\n");
        mavsdk_destroy(mavsdk);
        return 1;
    }
    
    // Create plugins
    mavsdk_action_t action = mavsdk_action_create(system);
    mavsdk_mission_t mission = mavsdk_mission_create(system);
    mavsdk_telemetry_t telemetry = mavsdk_telemetry_create(system);
    
    if (!action || !mission || !telemetry) {
        fprintf(stderr, "Failed to create plugins\n");
        if (action) mavsdk_action_destroy(action);
        if (mission) mavsdk_mission_destroy(mission);
        if (telemetry) mavsdk_telemetry_destroy(telemetry);
        mavsdk_destroy(mavsdk);
        return 1;
    }
    
    // Wait for system to be ready
    bool health_ok = false;
    while (!health_ok) {
        printf("Waiting for system to be ready\n");
        sleep(1);
        mavsdk_telemetry_health_t health;
        mavsdk_telemetry_health(telemetry, &health);
        health_ok = health.is_gyrometer_calibration_ok &&
                    health.is_accelerometer_calibration_ok &&
                    health.is_magnetometer_calibration_ok &&
                    health.is_local_position_ok &&
                    health.is_global_position_ok &&
                    health.is_home_position_ok;
    }
    
    printf("System ready\n");
    printf("Creating and uploading mission\n");
    
    // Create mission items array
    mavsdk_mission_mission_item_t* mission_items = 
        malloc(6 * sizeof(mavsdk_mission_mission_item_t));
    
    mission_items[0] = make_mission_item(
        47.398170327054473, 8.5456490218639658, 10.0f, 5.0f, false,
        20.0f, 60.0f, MAVSDK_MISSION_MISSION_ITEM_CAMERA_ACTION_NONE);
    
    mission_items[1] = make_mission_item(
        47.398241338125118, 8.5455360114574432, 10.0f, 2.0f, true,
        0.0f, -60.0f, MAVSDK_MISSION_MISSION_ITEM_CAMERA_ACTION_TAKE_PHOTO);
    
    mission_items[2] = make_mission_item(
        47.398139363821485, 8.5453846156597137, 10.0f, 5.0f, true,
        -45.0f, 0.0f, MAVSDK_MISSION_MISSION_ITEM_CAMERA_ACTION_START_VIDEO);
    
    mission_items[3] = make_mission_item(
        47.398058617228855, 8.5454618036746979, 10.0f, 2.0f, false,
        -90.0f, 30.0f, MAVSDK_MISSION_MISSION_ITEM_CAMERA_ACTION_STOP_VIDEO);
    
    mission_items[4] = make_mission_item(
        47.398100366082858, 8.5456969141960144, 10.0f, 5.0f, false,
        -45.0f, -30.0f, MAVSDK_MISSION_MISSION_ITEM_CAMERA_ACTION_START_PHOTO_INTERVAL);
    
    mission_items[5] = make_mission_item(
        47.398001890458097, 8.5455576181411743, 10.0f, 5.0f, false,
        0.0f, 0.0f, MAVSDK_MISSION_MISSION_ITEM_CAMERA_ACTION_STOP_PHOTO_INTERVAL);
    
    // Create mission plan
    mavsdk_mission_mission_plan_t mission_plan = {0};
    mission_plan.mission_items = mission_items;
    mission_plan.mission_items_size = 6;
    
    // Upload mission
    printf("Uploading mission...\n");
    mavsdk_mission_result_t upload_result = 
        mavsdk_mission_upload_mission(mission, mission_plan);
    
    if (upload_result != MAVSDK_MISSION_RESULT_SUCCESS) {
        fprintf(stderr, "Mission upload failed: %d, exiting.\n", upload_result);
        free(mission_items);
        mavsdk_mission_destroy(mission);
        mavsdk_action_destroy(action);
        mavsdk_telemetry_destroy(telemetry);
        mavsdk_destroy(mavsdk);
        return 1;
    }
    
    // Free the mission items array
    free(mission_items);
    
    // Arm
    printf("Arming...\n");
    mavsdk_action_result_t arm_result = mavsdk_action_arm(action);
    if (arm_result != MAVSDK_ACTION_RESULT_SUCCESS) {
        fprintf(stderr, "Arming failed: %d\n", arm_result);
        mavsdk_mission_destroy(mission);
        mavsdk_action_destroy(action);
        mavsdk_telemetry_destroy(telemetry);
        mavsdk_destroy(mavsdk);
        return 1;
    }
    printf("Armed.\n");
    
    // Subscribe to mission progress before starting
    mavsdk_mission_mission_progress_handle_t progress_handle = 
        mavsdk_mission_subscribe_mission_progress(
            mission, mission_progress_callback, NULL);
    
    // Start mission
    printf("Starting mission...\n");
    mavsdk_mission_result_t start_result = mavsdk_mission_start_mission(mission);
    if (start_result != MAVSDK_MISSION_RESULT_SUCCESS) {
        fprintf(stderr, "Starting mission failed: %d\n", start_result);
        mavsdk_mission_unsubscribe_mission_progress(mission, progress_handle);
        mavsdk_mission_destroy(mission);
        mavsdk_action_destroy(action);
        mavsdk_telemetry_destroy(telemetry);
        mavsdk_destroy(mavsdk);
        return 1;
    }
    
    // Wait for pause condition
    while (!want_to_pause) {
        sleep(1);
    }
    
    // Pause mission
    printf("Pausing mission...\n");
    mavsdk_mission_result_t pause_result = mavsdk_mission_pause_mission(mission);
    if (pause_result != MAVSDK_MISSION_RESULT_SUCCESS) {
        fprintf(stderr, "Failed to pause mission: %d\n", pause_result);
    }
    printf("Mission paused.\n");
    
    // Pause for 5 seconds
    sleep(5);
    
    // Continue mission
    printf("Continuing mission...\n");
    mavsdk_mission_result_t continue_result = mavsdk_mission_start_mission(mission);
    if (continue_result != MAVSDK_MISSION_RESULT_SUCCESS) {
        fprintf(stderr, "Starting mission again failed: %d\n", continue_result);
        mavsdk_mission_unsubscribe_mission_progress(mission, progress_handle);
        mavsdk_mission_destroy(mission);
        mavsdk_action_destroy(action);
        mavsdk_telemetry_destroy(telemetry);
        mavsdk_destroy(mavsdk);
        return 1;
    }
    
    // Wait for mission to finish
    bool is_finished = false;
    while (!is_finished) {
        sleep(1);
        mavsdk_mission_is_mission_finished(mission, &is_finished);
    }
    
    // Return to launch
    printf("Commanding RTL...\n");
    mavsdk_action_result_t rtl_result = mavsdk_action_return_to_launch(action);
    if (rtl_result != MAVSDK_ACTION_RESULT_SUCCESS) {
        fprintf(stderr, "Failed to command RTL: %d\n", rtl_result);
        mavsdk_mission_unsubscribe_mission_progress(mission, progress_handle);
        mavsdk_mission_destroy(mission);
        mavsdk_action_destroy(action);
        mavsdk_telemetry_destroy(telemetry);
        mavsdk_destroy(mavsdk);
        return 1;
    }
    printf("Commanded RTL.\n");
    
    // Wait a bit for armed state to be correct
    sleep(2);
    
    // Wait until disarmed
    bool armed = true;
    while (armed) {
        sleep(1);
        mavsdk_telemetry_armed(telemetry, &armed);
    }
    
    printf("Disarmed, exiting.\n");
    
    // Cleanup
    mavsdk_mission_unsubscribe_mission_progress(mission, progress_handle);
    mavsdk_mission_destroy(mission);
    mavsdk_action_destroy(action);
    mavsdk_telemetry_destroy(telemetry);
    mavsdk_destroy(mavsdk);
    
    return 0;
}

File: examples/mission.py
import time
import sys
from pymavsdk import *
from pymavsdk.plugins.action import *
from pymavsdk.plugins.mission import *
from pymavsdk.plugins.telemetry import *

def usage(bin_name):
    print(f"Usage : {bin_name} <connection_url>")
    print("Connection URL format should be :")
    print(" For TCP server: tcpin://<our_ip>:<port>")
    print(" For TCP client: tcpout://<remote_ip>:<port>")
    print(" For UDP server: udpin://<our_ip>:<port>")
    print(" For UDP client: udpout://<remote_ip>:<port>")
    print(" For Serial : serial://</path/to/serial/dev>:<baudrate>]")
    print("For example, to connect to the simulator use URL: udpin://0.0.0.0:14540")

def make_mission_item(
    latitude_deg,
    longitude_deg, 
    relative_altitude_m,
    speed_m_s,
    is_fly_through,
    gimbal_pitch_deg,
    gimbal_yaw_deg,
    camera_action
):
    """Create a mission item with the given parameters"""
    item = MissionItem()
    item.latitude_deg = latitude_deg
    item.longitude_deg = longitude_deg
    item.relative_altitude_m = relative_altitude_m
    item.speed_m_s = speed_m_s
    item.is_fly_through = is_fly_through
    item.gimbal_pitch_deg = gimbal_pitch_deg
    item.gimbal_yaw_deg = gimbal_yaw_deg
    item.camera_action = camera_action
    item.loiter_time_s = 0.0
    item.camera_photo_interval_s = 1.0
    item.acceptance_radius_m = 1.0
    item.yaw_deg = 0.0
    item.camera_photo_distance_m = 0.0
    item.vehicle_action = MissionItem.VehicleAction.NONE
    return item

def main():
    if len(sys.argv) != 2:
        usage(sys.argv[0])
        return 1

    configuration = Configuration.create_with_component_type(ComponentType.GROUND_STATION)
    mavsdk = Mavsdk(configuration)

    # Add connection
    connection_result = mavsdk.add_any_connection(sys.argv[1])

    if connection_result != ConnectionResult.SUCCESS:
        print(f"Connection failed: {connection_result}")
        return 1

    print("Waiting for system...")

    # Wait for autopilot system
    drone = None
    while drone is None:
        time.sleep(1)

        systems = mavsdk.get_systems()
        for system in systems:
            if system.has_autopilot() and system.is_connected():
                drone = system
                break

    print("Found autopilot system!")

    # Create plugins
    action = Action(drone)
    mission = Mission(drone)
    telemetry = Telemetry(drone)

    # Wait for system to be ready
    health_ok = False
    while not health_ok:
        print("Waiting for system to be ready")
        time.sleep(1)
        health = telemetry.health()
        health_ok = (health.is_gyrometer_calibration_ok and
                    health.is_accelerometer_calibration_ok and
                    health.is_magnetometer_calibration_ok and
                    health.is_local_position_ok and
                    health.is_global_position_ok and
                    health.is_home_position_ok)

    print("System ready")
    print("Creating and uploading mission")

    # Create mission items
    mission_items = [
        make_mission_item(
            47.398170327054473, 8.5456490218639658, 10.0, 5.0, False,
            20.0, 60.0, MissionItem.CameraAction.NONE
        ),
        make_mission_item(
            47.398241338125118, 8.5455360114574432, 10.0, 2.0, True,
            0.0, -60.0, MissionItem.CameraAction.TAKE_PHOTO
        ),
        make_mission_item(
            47.398139363821485, 8.5453846156597137, 10.0, 5.0, True,
            -45.0, 0.0, MissionItem.CameraAction.START_VIDEO
        ),
        make_mission_item(
            47.398058617228855, 8.5454618036746979, 10.0, 2.0, False,
            -90.0, 30.0, MissionItem.CameraAction.STOP_VIDEO
        ),
        make_mission_item(
            47.398100366082858, 8.5456969141960144, 10.0, 5.0, False,
            -45.0, -30.0, MissionItem.CameraAction.START_PHOTO_INTERVAL
        ),
        make_mission_item(
            47.398001890458097, 8.5455576181411743, 10.0, 5.0, False,
            0.0, 0.0, MissionItem.CameraAction.STOP_PHOTO_INTERVAL
        )
    ]

    # Create mission plan
    mission_plan = MissionPlan(mission_items)

    # Upload mission
    print("Uploading mission...")
    upload_result = mission.upload_mission(mission_plan)

    if upload_result != MissionResult.SUCCESS:
        print(f"Mission upload failed: {upload_result}, exiting.")
        return 1

    # Arm
    print("Arming...")
    arm_result = action.arm()
    if arm_result != ActionResult.SUCCESS:
        print(f"Arming failed: {arm_result}")
        return 1
    print("Armed.")

    # Subscribe to mission progress before starting
    want_to_pause = False

    def mission_progress_callback(progress, user_data=None):
        print(f"Mission status update: {progress.current} / {progress.total}")

        if progress.current >= 2:
            nonlocal want_to_pause
            want_to_pause = True

    progress_handle = mission.subscribe_mission_progress(mission_progress_callback)

    # Start mission
    print("Starting mission...")
    start_result = mission.start_mission()
    if start_result != MissionResult.SUCCESS:
        print(f"Starting mission failed: {start_result}")
        mission.unsubscribe_mission_progress(progress_handle)
        return 1

    # Wait for pause condition
    while not want_to_pause:
        time.sleep(1)

    # Pause mission
    print("Pausing mission...")
    pause_result = mission.pause_mission()
    if pause_result != MissionResult.SUCCESS:
        print(f"Failed to pause mission: {pause_result}")
    print("Mission paused.")

    # Pause for 5 seconds
    time.sleep(5)

    # Continue mission
    print("Continuing mission...")
    continue_result = mission.start_mission()
    if continue_result != MissionResult.SUCCESS:
        print(f"Starting mission again failed: {continue_result}")
        mission.unsubscribe_mission_progress(progress_handle)
        return 1

    # Wait for mission to finish
    is_finished = False
    while not is_finished:
        time.sleep(1)
        is_finished = mission.is_mission_finished()

    # Return to launch
    print("Commanding RTL...")
    rtl_result = action.return_to_launch()
    if rtl_result != ActionResult.SUCCESS:
        print(f"Failed to command RTL: {rtl_result}")
        mission.unsubscribe_mission_progress(progress_handle)
        return 1
    print("Commanded RTL.")

    # Wait a bit for armed state to be correct
    time.sleep(2)

    # Wait until disarmed
    armed = True
    while armed:
        time.sleep(1)
        armed = telemetry.armed()

    print("Disarmed, exiting.")

    # Cleanup
    mission.unsubscribe_mission_progress(progress_handle)

    return 0

if __name__ == "__main__":
    sys.exit(main())
